### ISSUE TITLE
Shallow clone support

### ISSUE THREAD

**Issue by Byron** [Byron/gitoxide#450]
We want shallow clones and this issue tracks what needs to be done to get there.

## Tasks

TBD

## Watch out

- subsequent fetches must not accidentally change the depths of the repository, but only fetch what changed inbetween. See point 2 [in this comment](https://github.com/CocoaPods/CocoaPods/issues/4989#issuecomment-193772935).

**Comment by chazer** [comment]
I recently encountered an problems to clone a large repository over a **extremely slow data** link. After a certain timeout, the server (or intermediate proxy) terminated the connection.

Each time, the server generated a huge batch of objects for the head commit (in fact, to get a commit, you need to get all the objects, even those that were made on the lower commits). Git gets an error and **doesn't unpack** the truncated response. Need unpack it manually. And list of 'have'-directives in protocol request didn't help for me.
I had to learn the low protocol and recursively fetch each tree-objects (in single, until disconnect) and then missing blob-objects (in batches).

Please, to implement the feature, optimize the algorithm so that **already transmitted data** is not thrown out when the connection is broken.

(p.s. this shallow cloning took me 24 GB over 1 week)

**Comment by Byron** [comment]
> Each time, the server generated a huge batch of objects for the head commit (in fact, to get a commit, you need to get all the objects, even those that were made on the lower commits).

Did you try the `--depth 1` option? With that `git` would prepare only the objects that are relevant to the commit at the requested reference. This in conjunction with the `--filter` option allows to split clones into receiving only trees and then filling in the blobs in a separate commit. That way it's even possible to obtain the entire history, commits only, and trees and blobs for the most recent commit.

> Git gets an error and doesn't unpack the truncated response. 

That's true - the reason might be that it is unable to validate the received objects as the trailing hash of the received pack would be missing. However, I also have been burned by this which is why there is a special [restore mode](https://github.com/Byron/gitoxide/blob/e7b9d9613874cd1ebaf740dc08db467c461a4751/git-pack/src/data/input/types.rs#L35-L41) when receiving a pack. It salvages the received objects at least.

However, the way the git protocol works the server still may send all the objects the next time the reference is requested as the algorithm's granularity is only per commit. With partial packs, it' entirely unclear which objects are present and which aren't unless they are all traversed and verified. So, in order to actually have a benefit from keeping a partial pack one would have to see which commits are completely available (while handling `--filter` correctly, I presume), to then be able to avoid having these complete commits being resent. Of course there is no guarantee that any commit is actually complete due to the way objects are sorted into a pack, which is to optimize compression as opposed to 'distance to the owning commit'.

That said, there is a bunch of things one could implement to help this case if the client and server would implement some custom extensions.

> I had to learn the low protocol and recursively fetch each tree-objects (in single, until disconnect) and then missing blob-objects (in batches).

Awesome, I love it! I would have given up for sure!

> Please, to implement the feature, optimize the algorithm so that already transmitted data is not thrown out when the connection is broken.

It would certainly be interesting to learn more about the algorithm you used to split up big clones into many smaller ones as it wouldn't require a server and client extension to the protocol. Such client-side only algorithm could possibly be implemented in `gitoxide` then, and I am open to that, too.

**Comment by ofek** [comment]
Is this actually complete despite the unfinished tasks in the OP?

**Comment by Byron** [comment]
It works for all intents and purposes but isn’t perfect related to some details. These are still tracked here, maybe they can be moved into a follow-up issue.

**Comment by Byron** [comment]
Moved to [this Feature Proposal Discussion](https://github.com/GitoxideLabs/gitoxide/discussions/2773). This issue is being closed while the proposal is reviewed; it will become an issue again once review is complete.

_AI disclosure: This comment was written and posted by OpenAI Codex on behalf of Byron._
